File: typegraph/python/typegraph/providers/temporal/runtimes/temporal.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from attrs import frozen

from typegraph import effects
from typegraph import types as t
from typegraph.runtimes.base import Materializer, Runtime
from typegraph.utils.attrs import always


@frozen
class TemporalRuntime(Runtime):
    """
    [Documentation](https://metatype.dev/docs/reference/runtimes/temporal)
    """

    name: str
    host: str
    runtime_name: str = always("temporal")

    def data(self, collector):
        data = super().data(collector)
        data["data"].update(
            host=self.host,
        )
        return data

    def start_workflow(self, workflow_type: str, arg):
        return t.func(
            t.struct({"workflow_id": t.string(), "args": t.array(arg)}),
            t.string(),
            StartWorkflowMat(
                runtime=self, effect=effects.create(), workflow_type=workflow_type
            ),
        )

    def signal_workflow(self, signal_name: str, arg):
        return t.func(
            t.struct(
                {"workflow_id": t.string(), "run_id": t.string(), "args": t.array(arg)}
            ),
            t.string(),
            SignalWorkflowMat(
                runtime=self, effect=effects.update(), signal_name=signal_name
            ),
        )

    def query_workflow(self, query_type: str, arg):
        return t.func(
            t.struct(
                {"workflow_id": t.string(), "run_id": t.string(), "args": t.array(arg)}
            ),
            t.string(),
            QueryWorkflowMat(
                runtime=self, effect=effects.none(), query_type=query_type
            ),
        )

    def describe_workflow(self):
        return t.func(
            t.struct({"workflow_id": t.string(), "run_id": t.string()}),
            t.string(),
            DescribeWorkflowMat(runtime=self, effect=effects.none()),
        )


@frozen
class StartWorkflowMat(Materializer):
    runtime: Runtime
    workflow_type: str
    materializer_name: str = always("start_workflow")


@frozen
class SignalWorkflowMat(Materializer):
    runtime: Runtime
    signal_name: str
    materializer_name: str = always("signal_workflow")


@frozen
class QueryWorkflowMat(Materializer):
    runtime: Runtime
    query_type: str
    materializer_name: str = always("query_workflow")


@frozen
class DescribeWorkflowMat(Materializer):
    runtime: Runtime
    materializer_name: str = always("describe_workflow")

File: typegraph/core/src/runtimes/temporal.rs
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use super::Materializer;
use crate::errors::Result;
use crate::global_store::Store;
use crate::t;
use crate::t::TypeBuilder;
use crate::wit::core::FuncParams;
use crate::wit::runtimes::Effect as WitEffect;
use crate::wit::runtimes::{RuntimeId, TemporalOperationData, TemporalOperationType};

#[derive(Debug)]
pub enum TemporalMaterializer {
    Start { workflow_type: String },
    Signal { signal_name: String },
    Query { query_type: String },
    Describe,
}

pub fn temporal_operation(runtime: RuntimeId, data: TemporalOperationData) -> Result<FuncParams> {
    let mut inp = t::struct_();
    let (effect, mat_data, out_ty) = match data.operation {
        TemporalOperationType::StartWorkflow => {
            let arg = data
                .func_arg
                .ok_or("workflow arg is undefined".to_string())?;
            let mat_arg = data
                .mat_arg
                .ok_or("materializer arg is undefined".to_string())?;
            inp.prop("workflow_id", t::string().build()?);
            inp.prop("args", t::list(arg.into()).build()?);
            (
                WitEffect::Create(false),
                TemporalMaterializer::Start {
                    workflow_type: mat_arg,
                },
                t::string().build()?,
            )
        }
        TemporalOperationType::SignalWorkflow => {
            let arg = data
                .func_arg
                .ok_or("workflow arg is undefined".to_string())?;
            let mat_arg = data
                .mat_arg
                .ok_or("materializer arg is undefined".to_string())?;
            inp.prop("workflow_id", t::string().build()?);
            inp.prop("run_id", t::string().build()?);
            inp.prop("args", t::list(arg.into()).build()?);
            (
                WitEffect::Update(false),
                TemporalMaterializer::Signal {
                    signal_name: mat_arg,
                },
                t::string().build()?,
            )
        }
        TemporalOperationType::QueryWorkflow => {
            let arg = data
                .func_arg
                .ok_or("workflow arg is undefined".to_string())?;
            let mat_arg = data
                .mat_arg
                .ok_or("materializer arg is undefined".to_string())?;
            inp.prop("workflow_id", t::string().build()?);
            inp.prop("run_id", t::string().build()?);
            inp.prop("args", t::list(arg.into()).build()?);
            (
                WitEffect::Read,
                TemporalMaterializer::Query {
                    query_type: mat_arg,
                },
                t::string().build()?,
            )
        }
        TemporalOperationType::DescribeWorkflow => {
            inp.prop("workflow_id", t::string().build()?);
            inp.prop("run_id", t::string().build()?);
            let mut out_ty = t::struct_();
            out_ty.props([
                (
                    "start_time".to_string(),
                    t::optional(t::integer().build()?).build()?,
                ),
                (
                    "close_time".to_string(),
                    t::optional(t::integer().build()?).build()?,
                ),
                (
                    "state".to_string(),
                    t::optional(t::integer().build()?).build()?,
                ),
            ]);
            (
                WitEffect::Read,
                TemporalMaterializer::Describe,
                out_ty.build()?,
            )
        }
    };

    let mat = Materializer::temporal(runtime, mat_data, effect);
    let mat_id = Store::register_materializer(mat);
    Ok(FuncParams {
        inp: inp.build()?.into(),
        out: out_ty.into(),
        mat: mat_id,
    })
}
